### ISSUE TITLE
Handle Clang-style jump tables

### ISSUE THREAD

**Issue by rohybrnta** [btbd/smap#1]
Firstly thanks for your great contribution, sadly I have an Issue I cannot solve my self. No matter what I try I always receive:

C:\smap\x64\Release>smap process.exe Test.dll
[-] searching for module alignments in 7720
[+] found 100 alignments

[-] analyzing sections...
[+] .text    > (0x1000, 0x111E00)

[error at C:\smap\translator.cpp:522] unexpected instruction with jump/offset register at 0000000000001BD7 (0000000000001BDE)

The target process is not protected.

**Comment by rohybrnta** [comment]
If I try it with your example.dll it prints up the following:
[error at C:\smap\translator.cpp:620] found invalid jump table entry (0000000000023045)

**Comment by btbd** [comment]
This is caused by an error in parsing a jump table. What compiler are you using for the DLL? Or if possible, can you upload your compiled output for the example DLL?

**Comment by rohybrnta** [comment]
For the first output I’ve used LLVM/clang, for the second one where I tried your example dll I’ve used msvc (vs 2019). I can upload the compiled example in about a hour

**Comment by btbd** [comment]
Compiling with clang I've found an edge case for jump tables where the case comparison occurred in an earlier branch. Pushed a fix (f2846192ac5241e2e62b44396fe4571b2253daf9), let me know if it fixes anything.

**Comment by rohybrnta** [comment]
The Issue seems to be still remaining:

C:\smap\x64\Release>smap Test.exe Test.dll
[-] searching for module alignments in 5836
[+] found 101 alignments

[-] analyzing sections...
[+] .text    > (0x1000, 0x111E00)

[error at C:\smap\translator.cpp:522] unexpected instruction with jump/offset register at 0000000000001BD7 (0000000000001BDE)

Is there any compiler options I need to take care off?

**Comment by btbd** [comment]
You don't have to change any compiler options, but it's hard for me to know what is unexpected without seeing the compiled output surrounding the RVA. Could you share the relevant assembly around that RVA?

**Comment by rohybrnta** [comment]
`.text:0000000000001BD7                 movsxd  rdx, ds:(jpt_1BDE - 1C84h)[rax+rcx*4]`
`text:0000000000001BDE                 jmp     rdx             ; switch jump`

**Comment by btbd** [comment]
So after some investigation, Clang may do jump tables like MSVC and other times like yours. With this type you see:
1. Clang is more direct and loads the jump table (1BCB) as the (1BD7) operand's base address instead of using a displacement.
2. Clang also uses `sub` instead of `cmp` for its JCC (1BB7)
3. Because of 2. Clang uses a different register than the index for its case comparison.

I've accounted for each of these in [this branch](https://github.com/btbd/smap/tree/clang) and did some tests. Let me know if there are any other cases I am missing with your DLL. Otherwise, I will merge.

**Comment by rohybrnta** [comment]
[-] analyzing sections...
[+] .text    > (0x1000, 0x111E00)

[error at C:\smap\translator.cpp:645] unexpected instruction (000000000006AA86, add r9d, 0x01) with index operand while parsing jump table (000000000006AAC4)

`.text:000000000006AA86                 add     r9d, 1          ; switch 6 cases
.text:000000000006AA8A                 mov     ecx, r9d
.text:000000000006AA8D                 sub     r9d, 5
.text:000000000006AA91                 mov     [rsp+128h+var_64], eax
.text:000000000006AA98                 mov     [rsp+128h+var_70], rcx
.text:000000000006AAA0                 mov     [rsp+128h+var_74], r9d
.text:000000000006AAA8                 ja      def_6AAC4       ; jumptable 000000000006AAC4 default case
.text:000000000006AAAE                 lea     rax, jpt_6AAC4
.text:000000000006AAB5                 mov     rcx, [rsp+128h+var_70]
.text:000000000006AABD                 movsxd  rdx, ds:(jpt_6AAC4 - 6ADECh)[rax+rcx*4]
.text:000000000006AAC1                 add     rdx, rax
.text:000000000006AAC4                 jmp     rdx             ; switch jump`

**Comment by btbd** [comment]
These edge cases are going to drive me crazy -- working on it now. 😄

**Comment by btbd** [comment]
@rohybrnta Simplified the logic so that after a JA, the first `sub` or `cmp` will override the current index search register (3504dd6ba50bdd6009ff2d023e11568ce31e5d25).

**Comment by rohybrnta** [comment]
This seems to do the job!

Is there any better way to load the imports as loading it myself using native inject?

`[error at C:\smap\translator.cpp:109] target process does not have export.dll loaded`
Using any other manual mapper works tho, any thoughts on this?
Thanks!

**Comment by rohybrnta** [comment]
I seem to have in issue with my imports using your manual mapper, other mmappers don't have these issues for me while resolving imports. Anything I can do to avoid that or improve that?
```
[error at C:\smap\translator.cpp:109] target process does not have MSVCP140.dll loaded
[error at C:\smap\translator.cpp:109] target process does not have d3dx9_43.dll loaded
[error at C:\smap\translator.cpp:109] target process does not have d3d9.dll loaded
[error at C:\smap\translator.cpp:109] target process does not have XINPUT9_1_0.dll loaded
[error at C:\smap\translator.cpp:109] target process does not have api-ms-win-crt-runtime-l1-1-0.dll loaded
```
Thanks

**Comment by btbd** [comment]
Standard manual mappers may manual map each dependency or try to standard inject it (if it's signed). But, you don't want your target process to require other dependencies. Compile your DLL statically linking all libraries.

**Comment by rohybrnta** [comment]
@btbd 
So run-time library should be /MD?
This was just an example of missing dependencies.

**Comment by btbd** [comment]
@rohybrnta /MT

**Comment by rohybrnta** [comment]
@btbd I compiled it using /MT when this issue happens

**Comment by btbd** [comment]
Is this DLL compiled with MSVC or Clang?

**Comment by rohybrnta** [comment]
Clang

**Comment by rohybrnta** [comment]
I looked at your example.dll linker and code generating options and set it in my project like yours.
Now I get the following output from your mmaper:
[error at C:\smap\translation.h:101] failed to generate valid assembly for modified translation
        idiv [rax]
        at 0000000000163869

The assembly associated to `0000000000163869`:

```
.text:0000000000163869                 idiv    cs:source_frequency
.text:0000000000163870                 mov     rcx, rax
.text:0000000000163873                 imul    rax, rdx, 3E8h
.text:000000000016387A                 imul    rcx, 3E8h

```

**Comment by btbd** [comment]
I don't use Clang but Google says `-static`. Not sure if this statically links all libraries or just the CRT.

**Comment by rohybrnta** [comment]
using -static seems to just link the CRT statically.

**Comment by btbd** [comment]
@rohybrnta I've changed the disassembler to always show the operand size (70109b28fef0e0f4610986b62e219867100599d2), which should solve the issue since the assembler would complain that `idiv [rax]` is ambiguous with no size i.e. `qword ptr [rax]`.

**Comment by btbd** [comment]
Closing this issue because the original problem is resolved (#2). Feel free to comment if you have any extra questions or open another issue if something is wrong.

**Comment by rohybrnta** [comment]
There still seems to be an issue with the jump table in the latest build:
`[error at C:\smap\translator.cpp:649] unexpected instruction (0000000000178CAE, lea r10d, [rdx+0x0F]) with index operand while parsing jump table (0000000000178CF7)`
